invenio_stats/processors.py

```diff
@@ -64,35 +65,27 @@ def anonymize_user(doc):
     salt = get_anonymization_salt(timestamp)
 
     visitor_id = hashlib.sha224(salt.encode('utf-8'))
+    unique_session_id = hashlib.sha224(salt.encode('utf-8'))
     # TODO: include random salt here, that changes once a day.
     # m.update(random_salt)
     if user_id:
         visitor_id.update(user_id.encode('utf-8'))
-    elif session_id:
-        visitor_id.update(session_id.encode('utf-8'))
-    elif ip and user_agent:
-        vid = '{}|{}|{}'.format(ip, user_agent, timeslice)
-        visitor_id.update(vid.encode('utf-8'))
-    else:
-        # TODO: add random data?
-        pass
-
-    unique_session_id = hashlib.sha224(salt.encode('utf-8'))
-    if user_id:
         sid = '{}|{}'.format(user_id, timeslice)
         unique_session_id.update(sid.encode('utf-8'))
     elif session_id:
+        visitor_id.update(session_id.encode('utf-8'))
         sid = '{}|{}'.format(session_id, timeslice)
         unique_session_id.update(sid.encode('utf-8'))
     elif ip and user_agent:
+        vid = '{}|{}'.format(ip, user_agent)
```

Comment:
What scares me is that there have been already 3-4 pairs of eyes over this logic (doesn't mean that they were correct though), which probably originally came from documentation found in other implementations, like e.g. [counter-processor](https://github.com/CDLUC3/counter-processor/blob/77c2c831995519990bc2ab3d98a92b0b9de5f14a/models/log_item.py#L56) and [its docs](https://github.com/CDLUC3/Make-Data-Count/blob/master/getting-started.md#counter-session-id-generation).

The [official Project COUNTER Code of Practice](https://www.projectcounter.org/code-of-practice-five-sections/7-processing-rules-underlying-counter-reporting-data/#doubleclick
) mentions IP address + User-Agent only. The [newest draft version](https://www.projectcounter.org/code-of-practice-rd-sections/7-processing-rules-underlying-reporting-data/#doubleclick_filtering) though, also mentions something about an hour-slice when describing doubleclick IDs...

Overall, what you're saying sounds correct, but I'm worried that this will change the computed aggregations in tests (and in existing deployments of `invenio-stats`). I think once the tests are fixed, we should review the numbers (and we could also give it a run with our tests on Zenodo, which also compare the resulting numbers)

Reply:
Can you please point to the doc? Thanks!

Reply:
Goddamit, I thought I made a comment but it glitched in GitHub's UI apparently... anyways here I go again:

I'm a bit scared about removing this one for the following reasons:

- There have been 3-4 pairs of eyes on the specific logic (of course this doesn't mean that they couldn't be wrong), and I think it's coming from already existing implementations by the Project COUNTER authors
- The [Code of Practice v5 on double-click filtering](https://www.projectcounter.org/code-of-practice-five-sections/7-processing-rules-underlying-counter-reporting-data/#doubleclick) mentions only IP Address + User-Agent (**this is the only differing source actually, but it's the official docs...**)
- On the contrary, [the newest release draft](https://www.projectcounter.org/code-of-practice-rd-sections/7-processing-rules-underlying-reporting-data/#doubleclick_filtering) mentions also something about an hour-slice
- One of the first and most popular implementations is the [CDLUC3/counter-processor](https://github.com/CDLUC3/counter-processor/blob/77c2c831995519990bc2ab3d98a92b0b9de5f14a/models/log_item.py#L56), which uses an hour timeslice for the doubleclick/visitor ID
- The [Make Data Count docs](https://github.com/CDLUC3/Make-Data-Count/blob/master/getting-started.md#counter-session-id-generation) mention also an hour timeslice

I think the practical thing here, is to see if the resulting numbers differ in the tests and make a decision after that (I'll also give it a go on Zenodo's side and see how/if the numbers change)

Reply:
I have reverted that change in my latest commit not for any of the aforementioned reasons, but to keep things simple.